accepted/2022/runtime-graph-maintainability.md

```diff
@@ -0,0 +1,60 @@
+
+Each time one of these distros ships a new version, we need to update our RID graph. If we don't, applications running on these new operating systems can fail because the RID is unknown to .NET. And it doesn't use the correct native assets.
+
+The underlying problem here is that RIDs were designed to be opaque strings. We intentionally didn't want people breaking these strings apart and inferring things about the RID. Instead, we build a [graph](https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.NETCore.Platforms/readme.md) that needs to be used in order to make sense of what RIDs are compatible with other RIDs.
+
+Each new version of Alpine gets a new RID value. `alpine.3.14`, `alpine.3.15`. When version `3.16` ships, `alpine.3.16` doesn't show up in the graph. .NET doesn't infer any compatibility about `alpine.3.16` and instead falls all the way back to `linux-{arch}` for the RID. This means any native asset built for `alpine` or `linux-musl` are no longer loaded causing apps to fail to execute properly.
+
+We spend a considerable amount of time just adding new versions to the RID graph in servicing. Here are some example PRs in the last 6 months:
+
+Alpine
+- 3.14
+    - [release/3.1](https://github.com/dotnet/corefx/pull/43099)
+    - [release/5.0](https://github.com/dotnet/runtime/pull/57226)
+    - [main / 6.0](https://github.com/dotnet/runtime/pull/55857)
+- 3.15
+    - [main](https://github.com/dotnet/runtime/pull/62938) Merged Jan 2022
+    - servicing hasn't happened yet, but will be needed for at least 3.1, 6.0, and possibly 5.0
+
+Fedora
+- 35
+    - [release/2.1](https://github.com/dotnet/corefx/pull/43033)
+    - [release/3.1](https://github.com/dotnet/corefx/pull/43032)
+    - [release/5.0](https://github.com/dotnet/runtime/pull/48203)
+    - [main / 6.0](https://github.com/dotnet/runtime/pull/48200)
+
+- 36
+    - [release/3.1](https://github.com/dotnet/corefx/pull/43101)
+    - [release/5.0](https://github.com/dotnet/runtime/pull/57703)
+    - [main / 6.0](https://github.com/dotnet/runtime/pull/57832)
+
+## Proposal
+
+Internally, we generate the full RID graph based on a higher-level set of data and use some rules for how to generate that data into the full RID graph. Here is the [documentation for that process](https://github.com/dotnet/runtime/tree/main/src/libraries/Microsoft.NETCore.Platforms#adding-a-new-os).
+
+In order to avoid having to service the `runtime.json` file for every new version, an approach we can take is the following:
+
+1. Add a new top-level section to the `runtime.json` JSON structure called `runtimeGroups`.
+   1. A runtime group will contain the following information:
+      1. `ID`: ex. `alpine`
+      2. `Parent`: ex. `linux-musl`
+      3. `TreatVersionsAsCompatible`: ex. `All`, `None`, `Major`, default is `All`
```

Comment:
How do we deal with sub-versions if any today? Eg., if There was a Fedora 36.1, or 36.0.1, would our infrastructure allow "fedora.36" assets - or require a "fedora.37"? If not, seems like that's another problem "All" would solve as it becomes unambiguous that 36.0.1 would be supported.